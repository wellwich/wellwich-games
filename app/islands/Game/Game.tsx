import { useEffect, useRef, useState } from 'hono/jsx';

const Game = ({ name }: { name: string }) => {
	const gameArea = useRef<HTMLDivElement>(null);
	const [gameTitle, setGameTitle] = useState('Game');
	let phaserInstance: Phaser.Game | null = null;

	useEffect(() => {
		const gameAreaElement = gameArea.current;
		if (!gameAreaElement) {
			console.error('gameAreaElement is null');
			return;
		}

		const makeGame = async () => {
			try {
				const phaser = await import(`../../games/${name}.ts`);
				phaserInstance = phaser.default(gameAreaElement);
				if (!phaserInstance) {
					console.error('phaserInstance is null');
					return;
				}
				setGameTitle(phaserInstance.config.gameTitle);

				const canvas = gameAreaElement.querySelector('canvas');
				if (canvas) {
					canvas.addEventListener('contextmenu', (e) => e.preventDefault());
				} else {
					console.error('Canvas element not found');
				}
			} catch (error) {
				console.error('Error loading Phaser game:', error);
			}
		};

		makeGame();

		return () => {
			if (phaserInstance) {
				phaserInstance.destroy(true);
				phaserInstance = null;
			}

			const canvas = gameAreaElement.querySelector('canvas');
			if (canvas) {
				canvas.removeEventListener('contextmenu', (e) => e.preventDefault());
			}
		};
	}, [name]);

	return (
		<div>
			<h2 class="text-2xl font-bold">{gameTitle}</h2>
			<div ref={gameArea} />
		</div>
	);
};

export default Game;
