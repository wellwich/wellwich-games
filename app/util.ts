function waitForElement(
	container: HTMLElement,
	selector: string,
): Promise<void> {
	return new Promise<void>((resolve) => {
		const observer = new MutationObserver((_mutations, obs) => {
			const element = container.querySelector(selector);
			if (element) {
				obs.disconnect();
				resolve();
			}
		});
		observer.observe(container, { childList: true, subtree: true });
	});
}

export { waitForElement };
