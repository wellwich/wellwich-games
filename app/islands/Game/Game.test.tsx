import { render } from 'hono/jsx/dom';
import { waitForElement } from '../../util';
import Game from './Game';

describe('ゲームコンポーネント', () => {
	it('レンダリングされるべき', async () => {
		const container = document.createElement('div');
		render(<Game name="test-game" />, container);

		await waitForElement(container, 'canvas');

		expect(container).toMatchSnapshot();
	});

	it('タイトルが正常に表示されるべき', async () => {
		const container = document.createElement('div');
		render(<Game name="test-game" />, container);

		await waitForElement(container, 'h2');

		const title = container.querySelector('h2');
		expect(title?.textContent).toBe('テストゲーム');
	});
});
