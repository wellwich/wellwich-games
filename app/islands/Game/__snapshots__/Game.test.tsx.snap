// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ゲームコンポーネント レンダリングされるべき 1`] = `
<div>
  <div
    class="mx-auto"
  >
    <h2
      class="text-2xl font-bold"
    >
      テストゲーム
    </h2>
    <div
      id="gameArea"
    >
      <canvas
        height="600"
        style="margin-left: 0px; margin-top: 0px;"
        width="800"
      />
    </div>
  </div>
</div>
`;
